pkg/machines/components/vmSnapshotsTab.jsx

```diff
@@ -43,118 +44,153 @@ function prettyTime(unixTime) {
-                        {snap.name}
-                    </div>
-                );
-            }
-        },
-        {
-            name: _("Description"), value: (snap, snapId) => {
-                let desc = snap.description;
-                if (!desc)
-                    desc = (<span className="snap-greyed-out">{_("No description")}</span>);
-
-                return (
-                    <div id={`${id}-snapshot-${snapId}-description`}>
-                        {desc}
-                    </div>
-                );
-            }
-        },
-        {
-            name: _("VM State"), value: (snap, snapId) => {
-                const statesMap = {
-                    shutoff: "shut off",
-                    "disk-snapshot": <span className="snap-greyed-out">{_("no state saved")}</span>,
-                };
-                const state = statesMap[snap.state] || snap.state;
-
-                const infoTips = {
-                    shutdown: _("Shutting down"),
-                    "disk-snapshot": _("Disk-only snapshot"),
-                    blocked: _("Domain is blocked on resource"),
-                    crashed: _("Domain has crashed"),
-                };
-                const tooltipMessage = infoTips[snap.state];
-                const tooltip = tooltipMessage
-                    ? (<span className="tooltip-circle">
-                        <Tooltip entryDelay={0} exitDelay={0} content={tooltipMessage}>
-                            <InfoAltIcon />
-                        </Tooltip>
-                    </span>) : null;
-
-                return (
-                    <div id={`${id}-snapshot-${snapId}-type`}>
-                        {state}
-                        {tooltip}
-                    </div>
-                );
-            }
-        },
-        {
-            name: _("Parent Snapshot"), value: (snap, snapId) => {
-                const parentName = snap.parentName || (<span className="snap-greyed-out">{_("No parent")}</span>);
-
-                return (
-                    <div id={`${id}-snapshot-${snapId}-parent`}>
-                        {parentName}
-                    </div>
-                );
-            }
-        },
-    ];
-
-    detailMap = detailMap.filter(d => !d.hidden);
-
-    const columnTitles = detailMap.map(target => target.name);
-    let rows = [];
-    if (vm.snapshots) {
-        rows = vm.snapshots.sort((a, b) => b.creationTime - a.creationTime).map((target, snapId) => {
-            const columns = detailMap.map(d => {
-                let column = null;
-                if (typeof d.value === 'string') {
-                    if (target[d.value] !== undefined)
-                        column = { title: <div id={`${id}-snapshot-${snapId}-${d.value}`}>{target[d.value]}</div> };
-                }
-                if (typeof d.value === 'function')
-                    column = { title: d.value(target, snapId) };
+class VmSnapshotsTab extends React.Component {
```

Comment:
How about using React Hooks (useState) and keeping this component functional?

This will make the diff nice mainly.

Reply:
The another PR https://github.com/cockpit-project/cockpit/pull/14283 also adds stateful logic into this class, so perhaps it makes sense to turn it into class once and for all.
But thanks, for the future, I will keep in my to not change functional components into classes because of few changes.